Qt: Implement Netplay

spycrab opened the pull request:
Ready to be reviewed and merged.

**Note: If DolphinQt crashes on startup, try starting it with the environment variable ``QT_NO_FT_CACHE`` set to ``1``.**

spycrab commented:
@dolphin-emu-bot rebuild

ligfx commented on Source/Core/DolphinQt2/GameList/GameList.cpp:
```diff
@@ -399,6 +401,24 @@ QString GameList::GetSelectedGame() const
+  else
+    view = m_list;
+
+  auto* model = view->model();
```
The get-model-from-view logic should be removed and replaced with simply `m_model`.

spycrab commented:
> GameFile::GetUniqueID(): since this is passed to NetPlay, the logic should be somewhere common so it stays in sync between DolphinWX and DolphinQt2

I'm afraid that would be rather complicated to accomplish as the Wx and Qt version of the function depend on the methods of ``ISOFile`` and ``GameFile`` respectively. If there's something I'm overlooking here feel free to tell me.

ligfx commented:
> > GameFile::GetUniqueID(): since this is passed to NetPlay, the logic should be somewhere common so it stays in sync between DolphinWX and DolphinQt2
>
> I'm afraid that would be rather complicated to accomplish as the Wx and Qt version of the function depend on the methods of ISOFile and GameFile respectively. If there's something I'm overlooking here feel free to tell me.

Can you use the methods from DiscIO::Volume?

spycrab commented:
Not without reparsing the Volume again, which I'd not be very fond of.

ligfx commented:
So don't reparse it, calculate and cache the Netplay identifier when the ISOFile/GameFile is first created, like the other properties.

spycrab commented:
You're right. Will move it to UICommon at some point...

spycrab commented:
> QtNetPlay::Init/GetGameList: Rather than passing around an arbitrary widget (that happens to contain the model we want), the GameListModel itself should be used directly. Maybe it should be pulled out to MainWindow level, and passed into the GameList widget?

Done.

BhaaLseN commented on Source/Core/DolphinQt2/DolphinQt2.vcxproj:
```diff
@@ -91,6 +91,8 @@
     <QtMoc Include="MainWindow.h" />
     <QtMoc Include="MenuBar.h" />
+    <QtMoc Include="NetPlay/NetPlayDialog.h" />
+    <QtMoc Include="NetPlay/NetPlaySetupDialog.h" />
```
Looks a little inconsistent with only this one having forward slashes instead of back slashes...

JosJuice commented:
Making a unified `GetUniqueIdentifier()` sounds like a bad idea to me. Since it won't be able to use the cached data from `GameFile`/`ISOFile`, you would waste time due to e.g. reading the names of Wii games twice, increasing game list scanning times. (The alternative of having different implementations in DolphinWX and DolphinQt2 isn't especially good either, but I can't think of anything better than that...)

ligfx commented on Source/Core/DolphinQt2/NetPlay/NetPlayDialog.cpp:
```diff
@@ -0,0 +1,549 @@
+
+  layout->addWidget(m_chat_edit, 0, 0, 1, -1);
+  layout->addWidget(m_chat_type_edit, 1, 0, 1, 4);
+  layout->addWidget(m_chat_send_button, 1, 4);
```
Should be

```c++
layout->addWidget(m_chat_type_edit, 1, 0);
layout->addWidget(m_chat_send_button, 1, 1);
```

If one of those needs to be stretched out and Qt isn't doing it automatically, either set it's sizePolicy or the QGridLayout's columnStretch.

ligfx commented on Source/Core/DolphinQt2/NetPlay/NetPlayDialog.cpp:
```diff
@@ -0,0 +1,549 @@
+  layout->addWidget(m_room_box, 0, 0);
+  layout->addWidget(m_hostcode_label, 0, 1);
+  layout->addWidget(m_hostcode_copy_button, 0, 2);
+  layout->addWidget(m_players_list, 1, 0, 1, 3);
```
```c++
layout->addWidget(m_players_list, 1, 0, 1, -1);
```

ligfx commented on Source/Core/DolphinQt2/NetPlay/NetPlayDialog.cpp:
```diff
@@ -0,0 +1,549 @@
+          !text.startsWith(QStringLiteral("Status:")))
+      {
+        item->setData(Qt::UserRole, player_ids[0]);
+        player_ids.erase(player_ids.begin());
```
Erasing from the beginning of a vector is expensive. Can't you increment an index into player_ids?

ligfx commented on Source/Core/DolphinQt2/NetPlay/NetPlayDialog.cpp:
```diff
@@ -0,0 +1,549 @@
+      m_hostcode_copy_button->setText(tr("Retry"));
+      m_hostcode_copy_button->setEnabled(true);
+      m_is_copy_button_retry = true;
+    }
```
missing `break;`

ligfx commented on Source/Core/DolphinQt2/NetPlay/PadMappingDialog.cpp:
```diff
@@ -0,0 +1,94 @@
+    m_gc_boxes[i] = new QComboBox;
+    m_wii_boxes[i] = new QComboBox;
+
+    m_main_layout->addWidget(new QLabel(tr("GC Pad %1").arg(i + 1)), 0, i);
```
*"GC Port %1"

ligfx commented on Source/Core/DolphinQt2/NetPlay/PadMappingDialog.cpp:
```diff
@@ -0,0 +1,94 @@
+    m_wii_boxes[i] = new QComboBox;
+
+    m_main_layout->addWidget(new QLabel(tr("GC Pad %1").arg(i + 1)), 0, i);
+    m_main_layout->addWidget(new QLabel(tr("Wiimote %1").arg(i + 1)), 0, 4 + i);
```
*"Wii Remote %1"

ligfx commented on Source/Core/DolphinQt2/NetPlay/PadMappingDialog.cpp:
```diff
@@ -0,0 +1,94 @@
+    m_main_layout->addWidget(m_wii_boxes[i], 1, 4 + i);
+  }
+
+  m_main_layout->addWidget(m_button_box, 2, 7);
```
Should be

```c++
m_main_layout->addWidget(m_button_box, 2, 0, 1, -1);
```

ligfx commented on Source/Core/DolphinQt2/NetPlay/PadMappingDialog.cpp:
```diff
@@ -0,0 +1,94 @@
+  for (const auto& player : m_players)
+    players.append(QString::fromStdString(player->name));
+
+  for (unsigned int i = 0; i < m_wii_boxes.size(); i++)
```
There's nothing in the code that guarantees m_wii_boxes and m_gc_boxes to be the same size. To be safe, and cause it's easy here, this is probably better:

```c++
for (auto& combo_group :{m_gc_boxes, m_wii_boxes})
{
  for (auto& combo : combo_group)
  {
    // ... 
  }
};
```

leoetlino commented:
Re config, loading/writing directly to an INI is (AFAIK) the standard way to handle config if you want to avoid stuffing everything into ConfigManager (which should be limited to Core settings). Should be replaced by the new system indeed.

mimimi085181 commented:
I can't test it, because it instant crashes after selecting qt from the nag screen:
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	DolphinQt2 pr-5810.exe
  Application Version:	0.0.0.0
  Application Timestamp:	59784418
  Fault Module Name:	Qt5Widgets.dll
  Fault Module Version:	5.9.0.0
  Fault Module Timestamp:	5928572b
  Exception Code:	c0000005
  Exception Offset:	000000000004b516
  OS Version:	6.1.7601.2.1.0.256.48
  Locale ID:	1031
  Additional Information 1:	7d33
  Additional Information 2:	7d336869c75c3c42b37b58551342880e
  Additional Information 3:	e79c
  Additional Information 4:	e79c5e4f99cd1f2f61da4311d417390a

spycrab commented:
I took the liberty of structuring this a bit to make it more readable.
### Done

> Trying to connect to NetPlay causes a segfault

I think I fixed that one shortly before you posted your review
> NetPlayDialog::UpdateGUI: Rather than using a timer, aren't we guaranteed to have Update called whenever any of this information changes?

Done.
### Questionable

> Why does it take forever to close the main NetPlay window? Sometimes it closes immediately, other times it takes seconds (and freezes the UI while doing so).

I'm guessing that's due to ``QtNetPlay::Quit()`` being time-consuming, but I'm not 100% sure.


> NetPlayDialog::OnStart: Should this "copy all relevant settings" logic be somewhere common? I don't know enough about NetPlay to know how important this logic is.

I'm not entirely sure... it does have some parameters that are GUI dependent.

### Going to change

>  NetPlayDialog: does this need to be a static instance? Can't it be owned by MainWindow?

Nope, it really doesn't need to. But I hadn't noticed that until you pointed it out. May also fix quite a lot of segfault issues since the Dialog is currently parentless.


> GameListModel is owned by GameList, but stored and passed around in places that aren't guaranteed to be within GameList's life time. (Besides, a widget should be focussed on user interaction, not mediating access to an internal model). The model should be instantiated someplace higher up (MainWindow? Settings?) and passed in to GameList.

Good idea, going on my TODO list.



> It's weird now to have two different singletons, Settings and QtNetPlay (and Settings isn't entirely just settings, see NANDRefresh). Maybe this is a good time to rename Settings, and then we can combine any global NetPlay stuff into that? I've been thinking of throwing it into a DolphinApplication class, since QApplications already have an instance() method and have guaranteed lifetime semantics (i.e. we create a destroy it in Main.cpp).

I would prefer to move it to the MainWindow, because that's where all NetPlay widgets will probably reside in. How should I go about passing those to  ``Settings`` (or whatever we decide to rename it to).

ligfx commented:
> I would prefer to move it to the MainWindow, because that's where all NetPlay widgets will probably reside in.

Oh yeah, if you can make something _not_ global, that's probably the better way to do it!

> How should I go about passing those to Settings (or whatever we decide to rename it to).

What do you mean?

spycrab commented:
> What do you mean?

Not relevant anymore. Also updated my response.

spycrab commented:
@ligfx Updated my response again.

ligfx commented:
Haven't had time to look at the code, but some quick observations from testing (all on macOS):
- Netplay main window on host: Clicking on game title pops up an empty game list
- Netplay main window: Layout on bottom row is wrong. Record inputs not be in a column with the player status box, and all widgets should be aligned left (not get extra space inserted between them)
- Netplay main window: Can't run md5 check twice (it gets greyed out)
- Netplay main window on host: For some reason both the chat entry and the game selection both get keyboard focus. Pressing enter to send a chat message also opens the game list window.
- Starting a netplay game and stopping the host causes the guest's game to freeze, with no indication that the host stopped the game.
- Host: start a netplay game, stop it, try to host again and the "Start" button is greyed out
- "Possible desync detected" message doesn't have arguments interpolated correctly

spycrab commented:
> Netplay main window on host: Clicking on game title pops up an empty game list

Fixed.
> Netplay main window: Can't run md5 check twice (it gets greyed out)

Fixed.
>Netplay main window on host: For some reason both the chat entry and the game selection both get keyboard focus. Pressing enter to send a chat message also opens the game list window.

Fixed.
> "Possible desync detected" message doesn't have arguments interpolated correctly

Fixed.
> Starting a netplay game and stopping the host causes the guest's game to freeze, with no indication that the host stopped the game.
> Host: start a netplay game, stop it, try to host again and the "Start" button is greyed out

Sadly those are general NetPlay issues according to @JMC47

>  Netplay main window: Layout on bottom row is wrong. Record inputs not be in a column with the player status box, and all widgets should be aligned left (not get extra space inserted between them)

Going to fix that.

ligfx commented on Source/Core/DolphinQt2/NetPlay/NetPlayDialog.cpp:
```diff
@@ -0,0 +1,568 @@
+    break;
+  case TraversalClient::VersionTooOld:
+    QMessageBox::critical(this, tr("Traversal Error"),
+                          tr("Dolphin is too old for traversal esrver"));
```
*"server"

spycrab commented:
@ligfx, @MayImilae: Should have adressed all your points (if I didn't overlook any...)


Except for this one:

> NetPlayDialog::reject / MainWindow::NetPlayQuit: nulling out the netplay pointers in both places is redundant.

No it isn't:  Nulling them will call their respective destructors and therefore Quit the NetPlayServer and Client respectively. If I wouldn't do that, the server / client would just keep running in the background.

ligfx commented:
> > NetPlayDialog::reject / MainWindow::NetPlayQuit: nulling out the netplay pointers in both places is redundant.
> 
> No it isn't: Nulling them will call their respective destructors and therefore Quit the NetPlayServer and Client respectively. If I wouldn't do that, the server / client would just keep running in the background.

Right, but you don't need to call the destructor twice in a row. It's called on both NetPlayDialog::reject and MainWindow::NetPlayQuit, which is connected to NetPlayDialog::reject. Only one of these needs to do it. In MainWindow is probably more appropriate, since that's also where the objects are instantiated.

ligfx commented:
Looking good, couple more things:

UI:
- MD5 dialog: The ordering of the players seems random (it changes every time I open the dialog). This should be sorted.
- MD5 dialog: DolphinWX says something like "Hashes match!" or "Hashes don't match!" which would be nice to have, rather than making people compare the hashes.
- If I host Netplay, start the game, then close the render window (through the window frame button, not by quitting Netplay), then the "Start" button becomes disabled until I restart Dolphin. DolphinWX lets me use the "Start" button again as soon as the emulation has stopped.
- NetPlayDialog::OnTraversalError: The default case should at least inform the user that _something_ bad happened with traversal, even if it doesn't specify what.

Code:
- The configuration logic needs to be updated to use the new-style configuration (available at Core/Config/NetplaySettings.h)
- MainWindow::NetPlayHost: this is the last remaining spot that creates a GameFile to read a Netplay ID: the NetplayID should be passed in directly from GameList and NetPlaySetupDialog, rather than the game path.
- NetPlayDialog::CreateMainLayout: The start button should be in the same HBoxLayout as the rest of the buttons on that row.
- NetPlayDialog::CreateMainLayout: The game button and chat box still take up five columns. They should only take up one column.
- And a number of threading things:
  - NetPlayDialog::OnMsgChangeGame: setting m_current_game needs to be done on the GUI thread
  - NetPlayDialog::OnMsgStartGame: m_current_game should be read from the GUI thread, and NetPlayClient::StartGame is expected to be called from the GUI thread
  - NetPlayDialog::OnPadBufferChanged: reading the value from m_buffer_size_box needs to be done on the GUI thread
  - NetPlayDialog::IsRecording: this needs to be run on the GUI thread. You can do something like `RunOnObject(this, [this]{ return m_record_input_box->isChecked(); });`
  - NetPlayDialog::FindGame: ditto. Use `RunOnObject`.

ligfx commented:
- GameList::NetPlayHost: The parameter is a netplay ID, not a path, and should be named appropriately (Unfortunate part of passing around strings is there's no type information to help figure out what a parameter means... I'd also be open to renaming the function to something like "NetPlayHostByUniqueID", though that's pretty verbose.).
- Why is NetPlayResetTraversalHost done in MainWindow? Can't this be done in NetPlaySetupDialog? That's also where the rest of the traversal config settings logic lives.
- MD5Dialog::SetProgress: Duplicated logic to get player name.
- MD5Dialog::SetProgress: This seems more complex than it needs to be. Why not set up all the widgets in show()? At that point, we know all the PIDs that will be used.
- NetPlayDialog::OnTraversalError: Please explicitly specify cases, so we get warnings if a new one is added and isn't handled.
- NetPlaySetupDialog::OnSaveSettings: The numeric settings here should be u16 (or uint16_t), which isn't necessarily the same thing as an unsigned short.

ligfx commented:
Sweet, the MD5Dialog stuff looks much simpler. One small code thing, then let's get this baby merged, it's been long enough:

- NetPlayDialog::OnTraversalError: Ah, I didn't mean that all cases should necessarily pop up a dialog box, but that the switch statement should handle all cases rather than using `default`. That way, the compiler will complain if a new enum case is added that isn't explicitly handled (well, it will, once #5906 lands). For example, in the latest change you forgot to handle VersionTooOld, which should raise a panic dialog or somehow inform the user.

  This should look something like:

   ```c++
  void NetPlayDialog::OnTraversalError(int error)
  {
    QueueOnObject(this, [this, error] {
      switch (error)
      {
      case TraversalClient::BadHost:
        QMessageBox::critical(this, tr("Traversal Error"), tr("Couldn't look up central server"));
        QDialog::reject();
        break;
      case TraversalClient::VersionTooOld:
        QMessageBox::critical(this, tr("Traversal Error"), tr("Dolphin is too old for traversal server"));
        QDialog::reject();
        break;
      case TraversalClient::ServerForgotAboutUs:
      case TraversalClient::SocketSendError:
      case TraversalClient::ResendTimeout:
        UpdateGUI();
        break;
      }
    });
  }
  ```
